### PR TITLE
Add frontend hook

### PR DESCRIPTION
**Pull request by Widcket**
### Description

This PR adds a frontend hook to fetch the User Profile from the React side of the app. The sample apps have been updated to use the provided hook.

### Testing

- [ ] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [ ] I have added documentation for new/changed functionality in this PR or in auth0.com/docs
- [ ] All active GitHub checks for tests, formatting, and security are passing
- [x] The correct base branch is being used, if not `master`

### CODE REVIEW

File: examples/api-call-example/components/with-auth.jsx [hunk 1]
```diff
@@ -1,42 +1,41 @@
...
+      const result = response.ok ? await response.json() : null;
+
+      return { user: result };
     }
 
-    render() {
-      if (!this.props.user) {
-        return <RedirectToLogin />;
-      }
+    const session = await auth0.getSession(ctx.req, ctx.res);
```
Comment (Widcket) [inline examples/api-call-example/components/with-auth.jsx:27]: The TS type for `getSession` only takes the request, but if the response is not passed it fails at runtime.

File: examples/typescript-example/components/with-auth.tsx [hunk 2]
```diff
@@ -1,48 +1,49 @@
...
-import { fetchUser } from '../lib/user';
 import createLoginUrl from '../lib/url-helper';
 import RedirectToLogin from '../components/login-redirect';
 
 type AuthenticatedProps = {
-  user?: any;
-  loading: boolean;
-  children: React.ReactChild | React.ReactChildren;
+  user?: UserProfile;
+  children: React.ReactNode;
```
Comment (adamjmcgrath) [inline examples/typescript-example/components/with-auth.tsx:11]: children should optional https://github.com/DefinitelyTyped/DefinitelyTyped/blob/4486794741ccb9f18f562fefade043ac31e817de/types/react/index.d.ts#L822

You could do `type AuthenticatedProps = React.PropsWithChildren & { user?: UserProfile }`

Which should resolve your issues on L32 and L45

File: src/hooks/use-user.tsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
+// eslint-disable-next-line import/no-extraneous-dependencies
+import React, { ReactElement, useState, useEffect, useContext, createContext } from 'react';
+
+declare const fetch: any;
```
Comment (adamjmcgrath) [inline src/hooks/use-user.tsx:4]: Can you add `lib: ['dom']` to the tsconfig, then you wont need to declare fetch https://github.com/auth0/auth0-react/blob/master/tsconfig.json#L4

**Comment by adamjmcgrath** [comment]
lgtm 👍 just a couple of TS things

We can look at the failing tests later

File: src/hooks/use-user.tsx (hunk 3 above)
Comment (adamjmcgrath) [reply src/hooks/use-user.tsx:4]: Or you may need to add something to eslint https://github.com/vercel/next.js/blob/canary/examples/with-typescript-eslint-jest/.eslintrc.json#L15

Comment (Widcket) [reply src/hooks/use-user.tsx]: Added `"browser": true` to eslint config. BTW, the fetch polyfill comes bundled with Next.js 9.4.0+. Currently the SDK requires 9.3.6. If we're going to keep that minimum version, we need to use `node-fetch` or similar here.

File: examples/typescript-example/components/with-auth.tsx (hunk 2 above)
Comment (Widcket) [reply examples/typescript-example/components/with-auth.tsx]: Changed in https://github.com/auth0/nextjs-auth0/pull/187/commits/5592ca5fe476d1148b8493c7feb81289e251fd15.

File: src/hooks/use-user.tsx (hunk 3 above)
Comment (adamjmcgrath) [reply src/hooks/use-user.tsx]: Let's go with 9.4 for min version

**Review by adamjmcgrath — APPROVED** [review]

**Pull request merged**